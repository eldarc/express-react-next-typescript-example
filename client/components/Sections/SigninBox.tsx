import * as React from 'react'
import { observer, inject } from 'mobx-react'
import Router from 'next/router'

// ~~~ Layout ~~~
import { Button, Form, FormGroup, Input, Label, Alert, Spinner } from 'reactstrap'

// ~~~ Interfaces ~~~
interface IProps {
  [x: string]: any
}

interface IState {
  mode: string
  processing: boolean
  error?: string
}

// ~~~ Component ~~~
@inject('store')
@observer
class SigninBox extends React.Component<IProps, IState> {
  private readonly username: React.RefObject<HTMLInputElement>
  private readonly password: React.RefObject<HTMLInputElement>

  constructor(props: IProps) {
    super(props)

    this.username = React.createRef()
    this.password = React.createRef()

    this.state = {
      mode: 'signin',
      processing: false,
      error: undefined
    }
  }

  public switchToSignin = () => {
    this.setState({
      mode: 'signin'
    })
  }

  public switchToSignup = () => {
    this.setState({
      mode: 'signup'
    })
  }

  public signin = async (event: React.FormEvent) => {
    event.preventDefault()

    try {
      await this.setState({
        processing: true
      })

      // @ts-ignore
      await this.props.store.auth.doSignin(this.username.current.value, this.password.current.value, this.state.mode === 'signup')
      await this.props.store.user.fetchUserData()
      await this.props.store.overview.ignoreUser(this.props.store.user.userData._id)

      await Router.push('/overview')
    } catch (e) {
      this.setState({
        error: JSON.stringify(e.response.data)
      })
    } finally {
      this.setState({
        processing: false
      })
    }
  }

  public render() {
    return (
        <div className="signin-box w-100">
          <div className="row">
            <div className="col-12 text-center mb-3">
              <h1>{this.state.mode === 'signin' ? 'Sign in' : 'Sign up'}</h1>
            </div>
            <div className="col-12 text-center mb-3">
              {
                this.state.mode === 'signin'
                  ? (
                      <span>
                        Don't have an account?&nbsp;
                        <a onClick={this.switchToSignup} href="#">Sign up here</a>
                      </span>
                    )
                  : (
                    <span>
                      Already have an account?&nbsp;
                      <a onClick={this.switchToSignin} href="#">Sign in here</a>
                    </span>
                  )
              }
            </div>
            <div className="col-12">
              { this.state.error ? (
                  <Alert color="danger">
                    There was an error while signing in: {this.state.error}
                  </Alert>
              ) : null}
              <Form onSubmit={this.signin}>
                <FormGroup>
                  <Label for="username">Username</Label>
                  <Input innerRef={this.username} type="text" name="username" id="username" placeholder="Please enter your username" />
                </FormGroup>
                <FormGroup>
                  <Label for="password">Password</Label>
                  <Input innerRef={this.password} type="password" name="password" id="password" placeholder="Please enter your password" />
                </FormGroup>

                <div className="d-flex w-100 justify-content-center">
                  <Button type="submit" color="primary">
                    {
                      this.state.processing ? <Spinner size="sm" color="light" />
                      : (this.state.mode === 'signin' ? 'Sign in' : 'Sign up')
                    }
                  </Button>
                </div>
              </Form>
            </div>
          </div>

          {/*language=SCSS*/}
          <style jsx>{`
              @import "variables";

              .signin-box {
                max-width: 450px;
                background-color: rgba($white, 0.4);
                border-radius: $border-radius-lg;
                padding: $grid-gutter-width / 2;
                box-shadow: 0 1px 2px 0 rgba(60,64,67,0.302),0 1px 3px 1px rgba(60,64,67,0.149);
              }
            `}</style>
        </div>
    )
  }
}

export default SigninBox
