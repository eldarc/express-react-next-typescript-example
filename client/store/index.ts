import { useStaticRendering } from 'mobx-react'
import AuthStore from './modules/AuthStore'
import UserStore from './modules/UserStore'
import OverviewStore from './modules/OverviewStore'
import Cookies from 'js-cookie'

// For SSR static rendering is required.
const isServer = (typeof window === 'undefined')
useStaticRendering(isServer)

// Main root store.
class Store {
  // @ts-ignore (AuthStore is definitely set in the `else` branch in the constructor)
  public auth: AuthStore
  public user: UserStore
  public overview: OverviewStore

  constructor(checkAuth: boolean) {
    this.overview = new OverviewStore()
    this.user = new UserStore()

    // Check if the user is already logged in, and create the store according to that status.
    const token = Cookies.get('token')
    if (checkAuth && token) {
      this.auth = new AuthStore(token)

      // Fetch user data.
      this.user.fetchUserData().then(() => {
        this.overview.ignoreUser(this.user.userData._id)
      })
    } else {
      this.auth = new AuthStore()
    }
  }
}

// Export Store and store initialization function.
export { Store }
let store: Store
export function initializeStore() {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    // return new Store(initialData)
    return new Store(false)
  }

  if (!store) {
    store = new Store(true)
  }

  return store
}
