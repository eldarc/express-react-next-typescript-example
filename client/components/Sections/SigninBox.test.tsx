/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { Store } from '@/store'

import SigninBox from './SigninBox'

describe('SigninBox', () => {
  it('renders correctly', () => {
    const tree = renderer
        .create(<SigninBox store={new Store(false)}/>)
        .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
