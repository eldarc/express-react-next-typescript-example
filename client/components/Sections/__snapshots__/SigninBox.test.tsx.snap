// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SigninBox renders correctly 1`] = `
<div
  className="jsx-1458856846 signin-box w-100"
>
  <div
    className="jsx-1458856846 row"
  >
    <div
      className="jsx-1458856846 col-12 text-center mb-3"
    >
      <h1
        className="jsx-1458856846"
      >
        Sign in
      </h1>
    </div>
    <div
      className="jsx-1458856846 col-12 text-center mb-3"
    >
      <span
        className="jsx-1458856846"
      >
        Don't have an account?Â 
        <a
          className="jsx-1458856846"
          href="#"
          onClick={[Function]}
        >
          Sign up here
        </a>
      </span>
    </div>
    <div
      className="jsx-1458856846 col-12"
    >
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className=""
            htmlFor="username"
          >
            Username
          </label>
          <input
            className="form-control"
            id="username"
            name="username"
            placeholder="Please enter your username"
            type="text"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            className=""
            htmlFor="password"
          >
            Password
          </label>
          <input
            className="form-control"
            id="password"
            name="password"
            placeholder="Please enter your password"
            type="password"
          />
        </div>
        <div
          className="jsx-1458856846 d-flex w-100 justify-content-center"
        >
          <button
            aria-label={null}
            className="btn btn-primary"
            onClick={[Function]}
            type="submit"
          >
            Sign in
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
